File: src/app/model/universe/physics/collision_logic/bodies_logic.c
#include <inttypes.h>

typedef struct {
    float x;
    float y;
} Vec2;

typedef struct {
    Vec2 position;
    float radius;
    float mass;
    Vec2 velocity;
} Body;

typedef struct {
    Vec2 delta1;
    Vec2 delta2;
} DeltaVelocities;

// return 0 if no collision
// return 1 if collision
uint8_t check_collision(Body body1, Body body2) {
    if (body2.position.x - body1.position.x < body1.radius) {
        return 1;
    }

    return 0;
}

// https://ru.wikipedia.org/wiki/Удар#Реальный_удар
// При коллизии.
// Нужно рассчитать насколько изменятся векторы скорости у тел.
// Предположительно, ты получишь новые скорости после вычислений, но
// от них нужно вычесть старые скорости и вернуть из функции.
// Если считать как в википедии, то пусть коэффициент восстановления
// будет как у стали: 0,55
DeltaVelocities calc_collision(Body body1, Body body2) {
    (void)body1;
    (void)body2;

    Vec2 delta1 = { .x = 0, .y = 0 };
    Vec2 delta2 = { .x = 100, .y = 100 };
    DeltaVelocities result = { .delta1 = delta1, .delta2 = delta2 };
    return result;
}

File: src/app/model/universe/physics/collision_logic/bodies_logic.rs
use super::Body;
use nannou::geom::Vec2;

pub struct BodiesLogic;

impl BodiesLogic {
    pub fn check_collision(&self, body1: &Body, body2: &Body) -> bool {
        unsafe {
            if check_collision(body1.clone(), body2.clone()) == 1 {
                true
            }
            else {
                false
            }
        }
    }

    pub fn calc_collision(&self, body1: &Body, body2: &Body) -> (Vec2, Vec2) {
        unsafe {
            let result = calc_collision(body1.clone(), body2.clone());
            (result.delta1, result.delta2)
        }
    }

    pub fn calc_collision_positions(&self, body1: &Body, body2: &Body) -> (Vec2, Vec2) {
        let radius_sum = body1.radius() + body2.radius();

        let distance_between_centers = Vec2::distance(body1.position(), body2.position());

        let delta = radius_sum - distance_between_centers;

        let delta_to_each = delta / 2.0;

        let position1 = body1.position() + (body1.position() - body2.position()).normalize_or_zero() * delta_to_each;
        let position2 = body2.position() + (body2.position() - body1.position()).normalize_or_zero() * delta_to_each;

        (position1, position2)
    }
}

#[repr(C)]
struct DeltaVelocities {
    delta1: Vec2,
    delta2: Vec2,
}

extern {
    fn check_collision(body1: Body , body2: Body) -> u8;

    fn calc_collision(body1: Body , body2: Body) -> DeltaVelocities;
}
